### PR TITLE
Add multi-fields support. Issue 84. Refactoring.

### LINKED ISSUE #84: in order to deal with multifield, try to get the docvalues field for retrieving the label
see https://github.com/elastic/kibana/pull/7302 for some example

### PR DESCRIPTION
**Pull request by SergeyBondarenko**
Closes #84 
Refactoring of the following PR: https://github.com/sirensolutions/kibi_timeline_vis/pull/97

### CODE REVIEW

File: public/kibi_timeline_vis_controller.js [hunk 1]
```diff
@@ -83,15 +83,21 @@ define(function (require) {
               searchSource.source(_.compact([ group.labelField, group.startField, group.endField ]));
               searchSource.set('filter', queryFilter.getFilters());
 
+              // save label field name to `fielddata_fields`
+              // it is needed to display not indexed fields in case of multi-fields
+              searchSource._state.fielddata_fields = _.compact([ group.labelField ]);
```
Comment (scampi) [inline public/kibi_timeline_vis_controller.js]: I think `compact` is not needed here.

File: public/kibi_timeline_vis_controller.js [hunk 2]
```diff
@@ -83,15 +83,21 @@ define(function (require) {
...
               $scope.visOptions.groups.push({
                 id: group.id,
                 color: group.color,
                 label: group.groupLabel,
                 searchSource: searchSource,
                 params: {
                   //kibi params
-                  labelFieldSequence: fields[i].byName[group.labelField].path,
-                  startFieldSequence: fields[i].byName[group.startField].path,
+                  labelFieldSequence: fields[i].byName[group.labelField].path.length > 0
```
Comment (scampi) [inline public/kibi_timeline_vis_controller.js]: why is this needed ?

File: public/lib/helpers/timeline_helper.js [hunk 3]
```diff
@@ -49,16 +49,12 @@ define(function (require) {
...
-        field = kibiUtils.getValuesAtPath(hit._source, params.labelFieldSequence);
-      } else {
-        field = _.get(hit._source, params.labelField);
-      }
-      if (field && (!_.isArray(field) || field.length)) {
-        return field;
-      }
-      return 'N/A';
+      // in kibi, we have the path property of a field
+      const field = params.labelFieldSequence ? params.labelFieldSequence[0] : params.labelField;
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: why `0` ?

File: public/lib/helpers/timeline_helper.js [hunk 4]
```diff
@@ -49,16 +49,12 @@ define(function (require) {
...
+      // get value from hit.fields in case of multi-fields
+      const value = _.get(hit._source, field) ? _.get(hit._source, field) : _.get(hit.fields, field);
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: you should use `kibiUtils.getValuesAtPath` instead of `_.get`. The reason is that `kibiUtils.getValuesAtPath` support object in array. For example with the following object:

```json
{
  "aaa": [
    {
      "bbb": 123
    }
  ]
}
```

To retrieve `123` you'd have to to call `_.get` with `[ 'aaa', '0', 'bbb' ]`, while with `kibiUtils.getValuesAtPath` you can just do `[ 'aaa', 'bbb' ]`.

---

With `hit.fields` however, just do `hit.fields[params.labelField]`.

File: public/kibi_timeline_vis_controller.js (hunk 2 above)
Comment (SergeyBondarenko) [inline public/kibi_timeline_vis_controller.js]: When you want a value from an ordinary (indexed) field, you can get it in this way `fields[i].byName[group.labelField].path`. But it returns an empty array in case of multi-fields. A multi-field field name you can get from `[group.labelField]`.

File: public/lib/helpers/timeline_helper.js (hunk 3 above)
Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: Because `params.labelFieldSequence` is an array with one element - field name. But maybe I can just return the array, I'll check it

File: public/lib/helpers/timeline_helper.js (hunk 4 above)
Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: ok

Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: thank you for the description

File: public/lib/helpers/timeline_helper.js (hunk 3 above)
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: it may have more than 1 element

Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: OK, for example, there we have two labels in one document: city.raw_0 and city.raw_1. Do I have to display values for both fields? If yes, they will be overlayed, one above other, because of the same date. 
Or you think about another case?

**Comment by jccq** [comment]
One below the other , one beside the other...

File: public/lib/helpers/timeline_helper.js [hunk 5]
```diff
@@ -75,6 +75,32 @@ define(function (require) {
     };
 
     /**
+     * pluckDate returns date field value/raw value
+     *
+     * @param hit the document of the event
+     * @param params configuration parameters for the event
```
Comment (szydan) [inline public/lib/helpers/timeline_helper.js]: There is no params 
instead there is field and fieldSequence 
please update the description

File: public/kibi_timeline_vis_controller.js [hunk 6]
```diff
@@ -95,11 +95,12 @@ define(function (require) {
                 params: {
                   //kibi params
                   // .path property doesn't exist in case of multi-fields
-                  labelFieldSequence: fields[i].byName[group.labelField].path.length === 0
+                  labelFieldSequence: !fields[i].byName[group.labelField].path.length
```
Comment (SergeyBondarenko) [inline public/kibi_timeline_vis_controller.js]: Use only `fields[i].byName[group.labelField].path` here. It is used to differentiate between Kibi and Kibana in timeline helper class methods.

File: public/kibi_timeline_vis_controller.js [hunk 7]
```diff
@@ -95,11 +95,12 @@ define(function (require) {
...
                     ? [ group.labelField ] : fields[i].byName[group.labelField].path,
-                  startFieldSequence: fields[i].byName[group.startField].path.length === 0
+                  startFieldSequence: !fields[i].byName[group.startField].path.length
```
Comment (SergeyBondarenko) [inline public/kibi_timeline_vis_controller.js]: Use only `fields[i].byName[group.startField].path` here. It is used to differentiate between Kibi and Kibana in timeline helper class methods.

File: public/kibi_timeline_vis_controller.js [hunk 8]
```diff
@@ -95,11 +95,12 @@ define(function (require) {
...
                     ? [ group.startField ] : fields[i].byName[group.startField].path,
-                  endFieldSequence: group.endField && fields[i].byName[group.endField].path || [],
+                  endFieldSequence: group.endField && !fields[i].byName[group.endField].path.length
```
Comment (SergeyBondarenko) [inline public/kibi_timeline_vis_controller.js]: Use only `group.endField && fields[i].byName[group.labelField].path || []` here. It is used to differentiate between Kibi and Kibana in timeline helper class methods.

File: public/lib/helpers/timeline_helper.js [hunk 9]
```diff
@@ -50,15 +61,66 @@ define(function (require) {
...
+        value = kibiUtils.getValuesAtPath(hit._source, field);
+        // get value from hit.fields in case of multi-fields
+        if (!value || !value.length) {
+          field = !field || !field.length ? params.labelField : field.join('.');
+          // '' if the field value is null
+          value = !hit.fields || !hit.fields[field] || hit.fields[field][0] === '' ? undefined : hit.fields[field];
+        }
       } else {
-        field = _.get(hit._source, params.labelField);
+        if (params.labelField) {
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: merge this `if` with the above `else`

File: public/lib/helpers/timeline_helper.js [hunk 10]
```diff
@@ -50,15 +61,66 @@ define(function (require) {
...
+          field = params.labelField;
+          value = _.get(hit._source, field);
+          if (!value) {
+            value = !hit.fields || !hit.fields[field] || hit.fields[field] === '' ? undefined : hit.fields[field];
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: create a function that does this and reuse it above

File: public/lib/helpers/timeline_helper.js [hunk 11]
```diff
@@ -50,15 +61,66 @@ define(function (require) {
...
+
+    /**
+     * pluckDate returns date field value/raw value
+     *
+     * @param hit the document of the event
+     * @param field to represent params.startField or params.endField
+     * @param fieldSequence to represent params.startFieldSequence or params.endFieldSequence
+     * @returns object with two properties: value and raw value
+     */
+    TimelineHelper.prototype.pluckDate = function (hit, field, fieldSequence) {
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: why is this needed ? take the text value of the date from `_source`, and its numeric value from the `fields`.

Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: Do you mean the `pluckDate` method? I just refactored [this duplication](https://github.com/sirensolutions/kibi_timeline_vis/pull/117/files/1af5594fd97fe867104d0e0dc4a549bade927789#diff-7419febb00f92050ffc6047cee9067a6L211) in the code. The code from line 211 till line 219 is a duplicate of the code above. 
As far as I understand `kibiUtils.getValuesAtPath()` is used to get value in case of Kibi. And `_.get(hit._source, params.endField)` - in case of Kibana.
Plus, I get multifields inside the `pluckDate` method.

Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: You suggest getting date values directly from `_source` and don't use `kibiUtils.getValuesAtPath()` and `_.get(hit._source, params.endField)`, don't you?

File: public/kibi_timeline.js [hunk 12]
```diff
@@ -184,10 +184,8 @@ define(function (require) {
           const events = [];
 
           if (params.startField) {
-            let startFieldValue;
-            let startRawFieldValue;
-            let endFieldValue;
-            let endRawFieldValue;
+            let startField = { value: undefined, raw: undefined };
+            let endField = { value: undefined, raw: undefined };
```
Comment (szydan) [inline public/kibi_timeline.js]: @scampi I would argue that leaving this as it is is more readable and I immediately know what this object should look like when populated
If this is not a problem lets leave it as it is

Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: I agree with @szydan suggestion about readability.

File: public/lib/helpers/timeline_helper.js [hunk 13]
```diff
@@ -50,15 +78,39 @@ define(function (require) {
...
-        field = kibiUtils.getValuesAtPath(hit._source, params.labelFieldSequence);
-      } else {
-        field = _.get(hit._source, params.labelField);
-      }
-      if (field && (!_.isArray(field) || field.length)) {
-        return field;
+      let value;
+
+      // in kibi, we have the path property of a field
+      if (params.labelFieldSequence || this.isMultiField(hit, params.labelField)) {
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: I think you can remove `isMultiField` completely and take care of the "multiField case of kibi" within the `else` block (there is nothing to change).
This way, you can remove line 89.

File: public/lib/helpers/timeline_helper.js [hunk 14]
```diff
@@ -41,6 +41,34 @@ define(function (require) {
...
+
+    /**
+     * getMultiFieldValue get a field value if the field is a multi-field
+     *
+     * @param hit the document of the event
+     * @param f field name
+     * @param style 'kibi'
+     * @returns field value
+     */
+    TimelineHelper.prototype.getMultiFieldValue = function (hit, f, style = false) {
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: Could this be simplified to:

```js
if (hit.fields && hit.fields[f] && hit.fields[f][0]) {
  return hit.fields[f];
}
```

**Comment by SergeyBondarenko** [comment]
@scampi I've done the suggested refactoring.

**Comment by scampi** [comment]
jenkins test it

**Comment by scampi** [comment]
@SergeyBondarenko please rebase

**Comment by SergeyBondarenko** [comment]
jenkins test it
